### PR TITLE
feat(curriculum): add permutation generator lab

### PR DESCRIPTION
**Pull request by zairahira**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Original prototype https://github.com/freeCodeCamp/CurriculumExpansion/pull/616/

I have modified the original code to remove redundant values from the final array.

### CODE REVIEW

**Comment by zairahira** [comment]
Reopening this PR as recursion is now part of the curriculum we'll ship.

File: curriculum/challenges/english/25-front-end-development/lab-permutation-generator/66fe4f33a2cc9b33f4d5cd9b.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+```js
+let emptyArr= permuteString('');
+
+// 1 because the empty string is being pushed and is a permutation of itself
+assert.lengthOf(emptyArr, 1); 
+
+```
+
+`permuteString('101')` should return `[ '101', '110', '011', '110', '101' ]`.
```
Comment (moT01) [inline curriculum/challenges/english/25-front-end-development/lab-permutation-generator/66fe4f33a2cc9b33f4d5cd9b.md:59]: This test is asking for duplicates - `101` and `110` are in there twice. The function should not return duplicates, correct? It also means that the solution that is passing this test isn't removing duplicates properly.

**Review by moT01 — CHANGES_REQUESTED** [review]
This is looking good @zairahira 👍 I left a comment about one of the tests and the solution below. Other than that, the only thing I saw was that you aren't actually required to use recursion to solve this. Wonder if you could add a test that checks if users are in fast using it?

**Comment by zairahira** [comment]
@moT01 
I have added this test to check for recursion.

````md

You should use recursion in your `permuteString` function.

```js
assert.match(permuteString.toString(), /permuteString\s*\(/);
```

````

Let me know what you think.

**Review by moT01 — APPROVED** [review]
LGTM 🎉

**Review by jdwilkin4 — APPROVED** [review]
